nexus/src/app/instance.rs

```diff
@@ -1311,6 +1313,31 @@ impl super::Nexus {
         )
         .await?;
 
+        // If the supplied instance state is at least as new as what's currently
+        // in the database, and it indicates the instance has no active VMM, the
+        // instance has been stopped and should have its virtual provisioning
+        // charges released.
+        //
+        // As with updating networking state, this must be done before
+        // committing the new runtime state to the database: once the DB is
+        // written, a new start saga can arrive and start the instance, which
+        // will try to create its own virtual provisioning charges, which will
+        // race with this operation.
+        if new_runtime_state.instance_state.propolis_id.is_none()
+            && new_runtime_state.instance_state.gen
+                >= db_instance.runtime().gen.0
+        {
+            self.db_datastore
```

Comment:
This check is made based off of `new_runtime_state` (from the sled agent) as well as `db_instance.runtime()` (from the DB), but both are cached -- couldn't **both** be out-of-date?

For example:

1. External API: Create instance, start it. Begin the call to "stop", which should send a request to stop down to the sled.
2. Sled Agent -> Nexus: Calls this function, indicating the instance should be stopped. Let's suppose sled agent thinks gen == N, the db thinks gen == N. But maybe calling `ensure_updated_instance_network_config` yields for a really long time.
3. Sled Agent -> Nexus: Calls this function again, from a retry-loop, but this time it succeeds. The instance is now stopped.
4. External API: Start the instance again. This re-provisions the resources used by the instance...
5. ... and, while the instance is running, the request from (2) comes back. It'll see a state of the world from before (3) and (4), so it'll call the `virtual_provisioning_collection_delete_instance` function. Now the accounting is wrong! This **is** a running instance, but we would have deleted the resources it's currently using.

Reply:
I think this is related to what I was touching on here: https://github.com/oxidecomputer/omicron/pull/4194#discussion_r1348088206

If we're associating these resources with VMMs (AKA, "what is actually running") rather than Instances (the opaque thing that exists even without a propolis), should we alter the calls to `virtual_provisioning_collection_{create, delete}_instance` to act on VMM UUIDs instead, rather than instance UUIDs?

That way, it wouldn't be possible to do a "double delete" -- once a VMM is gone, the UUID for it should never be re-used/re-added, so repeated/delayed requests to delete should have no further effect after the first successful call.

Reply:
Hmm. Let me think about this one a little more--the race you've identified is definitely a problem. I'll mull it over to see if there's a good way to resolve it without dipping into transaction-land (such that we could say "only proceed with the deletion if the instance generation number is less than the one I think I'm trying to apply" or suchlike).

It's true that VMMs don't share that problem because they can't come back from a terminal state, but I want to be careful about charging counters on a per-VMM basis if those counters are going to be used for quota management purposes (which AIUI is part of the discussion in RFD 427). For example, suppose I have a quota of 32 vCPUs on my project; I start an instance with 16 vCPUs; the rack operator starts an upgrade that migrates my instance; then I try to start a second instance with 16 vCPUs; if we track per-VMM, this will fail because my quota's exhausted--but what do you *mean* it's exhausted? I only have the one running instance! (cc @askfongjojo to help weigh in on this part; if we're OK with charging on a per-VMM basis even though this can cause transient double-counting, then switching to VMM IDs is a much simpler option here.)

Reply:
Fixed in 16b7fe6 by putting a sub-select into the instance provisioning counter deletion query that makes it apply only if the caller has furnished a sufficiently new generation number.

Reply:
Transient double-counting is probably fine if it's indeed transient (i.e. seconds, not minutes). But if we foresee failure modes that can cause in-progress migration to double count for an extended period of time, it'll be better for the system rather than the user to "absorb" the migration-in-transit usage.